File: src/store/blobs.rs
//! Content-addressed storage for the bytes an objective pins by hash.
//!
//! An objective commits to its checker by digest -- `evaluator_sha256`,
//! `checker_sha256` -- and those digests are part of the objective's id, so
//! editing a checker forks the objective rather than rescoring work already done
//! against it. That much has always been true. What was missing is anywhere to
//! *put* the bytes: the digest was only ever used to check a file the operator
//! already happened to have, at a relative path under
//! [`crate::verifiers::VerifierRegistry`]'s root.
//!
//! The consequence was narrow and awkward. Two nodes whose roots differ reach
//! different verdicts on the same claim -- one `Accept`, one `Unavailable` --
//! and while the rule that `Unavailable` never settles keeps that from being a
//! consensus failure, it means public verifiability depended on out-of-band file
//! distribution. "Anyone can re-derive every settled result from nothing but a
//! copy of the log" needed a copy of the verifier tree too.
//!
//! # The name is the hash
//!
//! Everything here follows from that. A blob is stored under its own digest, so:
//!
//! - **Verification is free and unavoidable.** [`BlobStore::get`] re-hashes what
//!   it read and refuses bytes that do not match the name they were filed under.
//!   There is no separate integrity record to keep in sync, because the filename
//!   *is* the integrity record.
//! - **Writes are idempotent.** Putting the same bytes twice is a no-op, and
//!   deliberately does not rewrite the file: modification time is what
//!   [`super::quota`] orders eviction by, so a redundant write would quietly
//!   promote a blob past older ones.
//! - **Two stores that hold the same blob hold the same bytes.** Which is what
//!   makes [`super::mirror`] a backup rather than a snapshot of one machine's
//!   idea of the truth.
//!
//! # Two decisions worth the paragraph
//!
//! **The `sha256:` prefix stays out of the filesystem.** Digests are written
//! `sha256:<64 hex>` everywhere else in this crate and that spelling is the API
//! here too, but `:` is not a portable filename character. The stored name is
//! the bare hex, so a store written on one platform is readable on another.
//!
//! **Blobs are sharded on the first byte** -- `blobs/ab/cdef...`. At the scale
//! this crate runs at today a flat directory would be fine, and that is exactly
//! why the layout has to be decided now: changing it later means moving every
//! file in every operator's store, and the cost of getting it right early is two
//! characters.
//!
//! # What this is not
//!
//! A fetch mechanism. This is where a blob lives once a node has it; nothing
//! here goes and gets one. That needs the gossip transport, which does not exist
//! -- `gossip.rs` is the merge law, not a wire protocol. Until it does, a blob
//! arrives by [`BlobStore::put`], by `proofwork blob put`, or by
//! [`super::mirror`] carrying a store that already had it.

use std::fs;
use std::path::{Path, PathBuf};

use super::{create_private_dir, io, StoreError};
use crate::canonical::{digest_bytes, DIGEST_PREFIX};

/// Subdirectory of the cache holding content-addressed blobs.
pub const BLOB_DIR: &str = "blobs";

/// Characters of hex used as the shard directory.
const SHARD: usize = 2;
/// Hex characters in a SHA-256 digest.
const HEX_LEN: usize = 64;

/// A content-addressed store: bytes filed under their own SHA-256.
///
/// Cheap to construct and holds no handles, so a caller that wants one per
/// command is not paying for it.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BlobStore {
    root: PathBuf,
}

impl BlobStore {
    /// A blob store rooted at `root`, which is the directory the shards live in.
    pub fn new(root: impl Into<PathBuf>) -> BlobStore {
        BlobStore { root: root.into() }
    }

    /// The blob store belonging to a data directory: `<data-dir>/cache/blobs`.
    ///
    /// Under `cache/` because a blob is by definition re-fetchable -- it is
    /// named by its content, so a lost blob is a re-download and never a lost
    /// record. That classification is not the whole story once an objective
    /// depends on one; see [`super::Store::with_pinned_blobs`].
    pub fn under(store: &super::Store) -> BlobStore {
        BlobStore::new(store.blob_dir())
    }

    pub fn root(&self) -> &Path {
        &self.root
    }

    /// Create the directory, owner-only where the platform allows it.
    pub fn prepare(&self) -> Result<(), StoreError> {
        create_private_dir(&self.root)
    }

    /// Where `digest` would live, or an error if it is not a digest.
    ///
    /// Returns a path whether or not anything is there. Callers that want the
    /// question answered use [`BlobStore::has`].
    pub fn path_of(&self, digest: &str) -> Result<PathBuf, StoreError> {
        let hex = hex_of(digest)?;
        let (shard, rest) = hex.split_at(SHARD);
        Ok(self.root.join(shard).join(rest))
    }

    /// Whether this store holds `digest`.
    ///
    /// Existence only. A file that is present but corrupt answers `true` here
    /// and fails in [`BlobStore::get`], which is the right split: this is a
    /// question about the store's inventory, and the integrity check belongs on
    /// the path that is about to use the bytes.
    pub fn has(&self, digest: &str) -> bool {
        matches!(self.path_of(digest), Ok(path) if path.is_file())
    }

    /// File `bytes` under their digest and return it.
    ///
    /// Idempotent: bytes already present are left exactly as they are, mtime
    /// included.
    pub fn put(&self, bytes: &[u8]) -> Result<String, StoreError> {
        let digest = digest_bytes(bytes);
        let path = self.path_of(&digest)?;
        if path.is_file() {
            return Ok(digest);
        }
        let parent = path.parent().unwrap_or(&self.root);
        create_private_dir(parent)?;

        // Write beside the destination and rename onto it. Same directory, so
        // the rename is atomic on any filesystem worth using, and a crash
        // halfway through leaves a `.tmp-` file rather than a truncated blob
        // sitting under a digest that promises otherwise. Everything downstream
        // trusts the name, so a partially written blob is worse than no blob.
        let temporary = parent.join(format!(".tmp-{}-{}", std::process::id(), &digest[..12]));
        fs::write(&temporary, bytes)
            .map_err(|e| io(format!("writing {}", temporary.display()), e))?;
        match fs::rename(&temporary, &path) {
            Ok(()) => Ok(digest),
            Err(error) => {
                let _ = fs::remove_file(&temporary);
                Err(io(format!("filing {}", path.display()), error))
            }
        }
    }

    /// Read a file and file its contents.
    pub fn put_file(&self, source: &Path) -> Result<String, StoreError> {
        let bytes = fs::read(source).map_err(|e| io(format!("reading {}", source.display()), e))?;
        self.put(&bytes)
    }

    /// Fetch `digest`, verifying that what came back hashes to it.
    ///
    /// `Ok(None)` means the store does not have it, which is a routine answer
    /// and not an error -- the caller falls back or reports `Unavailable`.
    /// [`StoreError::CorruptBlob`] means it *did* have it and the bytes are
    /// wrong, which is a fact about this disk and never about the objective that
    /// asked for them.
    pub fn get(&self, digest: &str) -> Result<Option<Vec<u8>>, StoreError> {
        let path = self.path_of(digest)?;
        let bytes = match fs::read(&path) {
            Ok(bytes) => bytes,
            Err(error) if error.kind() == std::io::ErrorKind::NotFound => return Ok(None),
            Err(error) => return Err(io(format!("reading {}", path.display()), error)),
        };
        let actual = digest_bytes(&bytes);
        if actual != canonical_digest(digest)? {
            return Err(StoreError::CorruptBlob {
                digest: canonical_digest(digest)?,
                actual,
            });
        }
        Ok(Some(bytes))
    }

    /// Every blob held, as `(digest, bytes)`, in digest order.
    ///
    /// Names that are not digests are skipped rather than reported: a `.tmp-`
    /// file from an interrupted [`BlobStore::put`] is debris, and an inventory
    /// that refused to run because somebody dropped a note in the directory
    /// would be useless exactly when it is needed.
    pub fn list(&self) -> Result<Vec<(String, u64)>, StoreError> {
        let mut blobs = Vec::new();
        let shards = match fs::read_dir(&self.root) {
            Ok(shards) => shards,
            Err(error) if error.kind() == std::io::ErrorKind::NotFound => return Ok(blobs),
            Err(error) => return Err(io(format!("reading {}", self.root.display()), error)),
        };
        for shard in shards {
            let shard = shard.map_err(|e| io(format!("reading {}", self.root.display()), e))?;
            if !shard.file_type().map(|t| t.is_dir()).unwrap_or(false) {
                continue;
            }
            let prefix = shard.file_name().to_string_lossy().into_owned();
            if prefix.len() != SHARD || !is_hex(&prefix) {
                continue;
            }
            let entries = fs::read_dir(shard.path())
                .map_err(|e| io(format!("reading {}", shard.path().display()), e))?;
            for entry in entries {
                let entry =
                    entry.map_err(|e| io(format!("reading {}", shard.path().display()), e))?;
                let rest = entry.file_name().to_string_lossy().into_owned();
                if rest.len() != HEX_LEN - SHARD || !is_hex(&rest) {
                    continue;
                }
                let bytes = entry.metadata().map(|m| m.len()).unwrap_or(0);
                blobs.push((format!("{DIGEST_PREFIX}{prefix}{rest}"), bytes));
            }
        }
        blobs.sort();
        Ok(blobs)
    }

    /// Re-hash everything held and return the digests whose bytes disagree.
    ///
    /// "The name is the hash" is the store's entire trust model, so it is worth
    /// being able to check rather than assume. An empty result is the good one.
    pub fn verify(&self) -> Result<Vec<String>, StoreError> {
        let mut corrupt = Vec::new();
        for (digest, _) in self.list()? {
            match self.get(&digest) {
                Ok(_) => {}
                Err(StoreError::CorruptBlob { digest, .. }) => corrupt.push(digest),
                Err(other) => return Err(other),
            }
        }
        Ok(corrupt)
    }
}

/// Accept a digest in either spelling and return the canonical, prefixed one.
///
/// Public because a pin set is a set of digests that has to compare equal
/// however its members were spelled: an objective's bare-hex `evaluator_sha256`
/// and a prefixed record id name the same blob, and a set that held both would
/// pin one and evict the other. See [`super::Store::with_pinned_blobs`].
pub fn canonical_digest(digest: &str) -> Result<String, StoreError> {
    Ok(format!("{DIGEST_PREFIX}{}", hex_of(digest)?))
}

/// The 64 hex characters of a digest, with or without the `sha256:` prefix.
///
/// Both spellings are accepted because both occur honestly: objectives write
/// `evaluator_sha256` as bare hex, while record ids and
/// [`crate::canonical::digest_bytes`] carry the prefix. Refusing one of them
/// would push a `strip_prefix` into every caller, and a caller that forgot it
/// would look up a blob that can never be there.
fn hex_of(digest: &str) -> Result<String, StoreError> {
    let hex = digest.strip_prefix(DIGEST_PREFIX).unwrap_or(digest);
    if hex.len() != HEX_LEN || !is_hex(hex) {
        return Err(StoreError::NotADigest(digest.to_string()));
    }
    Ok(hex.to_ascii_lowercase())
}

/// Lowercase hex only. Uppercase is refused rather than accepted and folded,
/// because two spellings of one digest are two filenames for one blob.
fn is_hex(text: &str) -> bool {
    !text.is_empty()
        && text
            .bytes()
            .all(|b| b.is_ascii_digit() || (b'a'..=b'f').contains(&b))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::store::tests::scratch;
    use crate::store::Store;

    fn store(tag: &str) -> (PathBuf, BlobStore) {
        let dir = scratch(tag);
        let blobs = BlobStore::new(dir.join("blobs"));
        blobs.prepare().expect("prepares");
        (dir, blobs)
    }

    #[test]
    fn a_blob_is_filed_under_its_own_digest() {
        let (dir, blobs) = store("put");
        let digest = blobs.put(b"print('hello')").expect("puts");
        assert_eq!(digest, digest_bytes(b"print('hello')"));
        assert!(blobs.has(&digest));
        assert_eq!(
            blobs.get(&digest).expect("gets"),
            Some(b"print('hello')".to_vec())
        );

        // Sharded on the first byte, and the `sha256:` prefix stays out of the
        // filesystem so the layout is portable.
        let hex = digest.strip_prefix(DIGEST_PREFIX).expect("prefixed");
        let path = blobs.path_of(&digest).expect("a digest");
        assert_eq!(path, blobs.root().join(&hex[..2]).join(&hex[2..]));
        assert!(!path.to_string_lossy().contains(':'));
        let _ = fs::remove_dir_all(&dir);
    }

    #[test]
    fn putting_the_same_bytes_twice_does_not_rewrite_the_file() {
        // Not merely an optimisation. `quota` evicts oldest-first by mtime, so a
        // redundant write would silently promote a blob past older ones and
        // change which content survives a reclaim.
        let (dir, blobs) = store("idempotent");
        let digest = blobs.put(b"same").expect("puts");
        let path = blobs.path_of(&digest).expect("a digest");
        let first = fs::metadata(&path).expect("metadata").modified().ok();

        let again = blobs.put(b"same").expect("puts again");
        assert_eq!(again, digest);
        let second = fs::metadata(&path).expect("metadata").modified().ok();
        assert_eq!(first, second, "an existing blob is left alone");
        assert_eq!(blobs.list().expect("lists").len(), 1);
        let _ = fs::remove_dir_all(&dir);
    }

    #[test]
    fn corrupt_bytes_are_refused_rather_than_returned() {
        // The whole trust model in one test: the filename is the integrity
        // record, so tampering is detected without any second copy of anything.
        let (dir, blobs) = store("corrupt");
        let digest = blobs.put(b"honest").expect("puts");
        let path = blobs.path_of(&digest).expect("a digest");
        fs::write(&path, b"tampered").expect("overwrite");

        assert!(blobs.has(&digest), "still on disk");
        match blobs
            .get(&digest)
            .expect_err("must not hand back the bytes")
        {
            StoreError::CorruptBlob {
                digest: reported,
                actual,
            } => {
                assert_eq!(reported, digest);
                assert_eq!(actual, digest_bytes(b"tampered"));
            }
            other => panic!("expected CorruptBlob, got {other:?}"),
        }
        assert_eq!(blobs.verify().expect("verifies"), vec![digest]);
        let _ = fs::remove_dir_all(&dir);
    }

    #[test]
    fn a_missing_blob_is_not_an_error() {
        // The routine answer. A caller that gets `None` falls back to the path,
        // and only then reports `Unavailable` -- an absent blob is a fact about
        // this node, and turning it into an error here would push that decision
        // to the wrong layer.
        let (dir, blobs) = store("missing");
        let absent = digest_bytes(b"never stored");
        assert!(!blobs.has(&absent));
        assert_eq!(blobs.get(&absent).expect("no error"), None);
        let _ = fs::remove_dir_all(&dir);
    }

    #[test]
    fn both_spellings_of_a_digest_find_the_same_blob() {
        // Objectives write `evaluator_sha256` as bare hex; ids and
        // `digest_bytes` carry the prefix. Accepting one only would mean a
        // caller that forgot to strip it looks up a blob that can never exist.
        let (dir, blobs) = store("spelling");
        let digest = blobs.put(b"either way").expect("puts");
        let bare = digest.strip_prefix(DIGEST_PREFIX).expect("prefixed");
        assert!(blobs.has(bare));
        assert_eq!(
            blobs.path_of(bare).expect("a digest"),
            blobs.path_of(&digest).expect("a digest")
        );
        let _ = fs::remove_dir_all(&dir);
    }

    #[test]
    fn a_name_that_is_not_a_digest_is_refused() {
        let (dir, blobs) = store("notadigest");
        for bad in [
            "",
            "sha256:",
            "beef",
            "sha256:not-hex-at-all-not-hex-at-all-not-hex-at-all-not-hex-at-all-xx",
            // Uppercase is refused rather than folded: two spellings of one
            // digest would be two filenames for one blob.
            "sha256:ABCDEF0123456789abcdef0123456789abcdef0123456789abcdef0123456789",
        ] {
            assert!(
                matches!(blobs.path_of(bad), Err(StoreError::NotADigest(_))),
                "{bad:?} should not name a blob"
            );
            assert!(!blobs.has(bad));
        }
        let _ = fs::remove_dir_all(&dir);
    }

    #[test]
    fn the_inventory_skips_debris_and_sorts() {
        let (dir, blobs) = store("list");
        let one = blobs.put(b"a").expect("puts");
        let two = blobs.put(b"bb").expect("puts");

        // Debris an interrupted `put` or a curious operator could leave behind.
        fs::write(blobs.root().join("README"), b"notes").expect("write");
        fs::create_dir_all(blobs.root().join("zz")).expect("mkdir");
        fs::write(blobs.root().join("zz").join("stray"), b"x").expect("write");
        let shard = blobs
            .path_of(&one)
            .expect("a digest")
            .parent()
            .expect("has a parent")
            .to_path_buf();
        fs::write(shard.join(".tmp-1234-abcdef012345"), b"partial").expect("write");

        let listed = blobs.list().expect("lists");
        let digests: Vec<&str> = listed.iter().map(|(d, _)| d.as_str()).collect();
        let mut expected = vec![one.as_str(), two.as_str()];
        expected.sort();
        assert_eq!(digests, expected);
        for (digest, bytes) in &listed {
            assert_eq!(*bytes, if *digest == one { 1 } else { 2 });
        }
        let _ = fs::remove_dir_all(&dir);
    }

    #[test]
    fn an_absent_store_lists_empty_rather_than_failing() {
        let dir = scratch("absent");
        let blobs = BlobStore::new(dir.join("never-created"));
        assert_eq!(blobs.list().expect("lists"), Vec::new());
        assert_eq!(blobs.verify().expect("verifies"), Vec::<String>::new());
        let _ = fs::remove_dir_all(&dir);
    }

    #[test]
    fn a_store_puts_its_blobs_under_the_cache() {
        // Where the bytes live decides whether the quota may delete them, so the
        // location is a contract rather than a detail.
        let dir = scratch("under");
        let store = Store::new(dir.join("data"));
        store.prepare().expect("prepares");
        let blobs = BlobStore::under(&store);
        assert_eq!(blobs.root(), store.cache_dir().join(BLOB_DIR));
        let digest = blobs.put(b"cached").expect("puts");
        assert_eq!(
            store.classify(&blobs.path_of(&digest).expect("a digest")),
            crate::store::Class::Reclaimable,
            "an unreferenced blob is a copy of something, so it may go"
        );
        let _ = fs::remove_dir_all(&dir);
    }

    #[test]
    fn put_file_and_put_agree() {
        let (dir, blobs) = store("putfile");
        let source = dir.join("checker.py");
        fs::write(&source, b"def score(a):\n    return 0\n").expect("write");
        let from_file = blobs.put_file(&source).expect("puts");
        let from_bytes = blobs.put(b"def score(a):\n    return 0\n").expect("puts");
        assert_eq!(from_file, from_bytes);
        let _ = fs::remove_dir_all(&dir);
    }
}

File: reference/python/proofwork/blobs.py
"""Content-addressed storage for pinned verifier code.

An objective pins its checker by SHA-256 and *references* it by a path relative
to the bundle root. The hash makes substitution impossible; the path makes
availability somebody else's problem. A node holding the log but not the
operator's working directory cannot run the checker, so it answers UNAVAILABLE
forever -- which means the guarantee the whole project rests on, *anyone can
independently re-derive every settled result from the log alone*, held only for
whoever also had the filesystem.

THE PIN IS ALREADY THE CONTENT ADDRESS

So there is no second identity scheme and, deliberately, no new record field. A
blob's name *is* the hex SHA-256 the objective already declares in
``checker_sha256`` / ``evaluator_sha256`` / ``statistic.sha256``. Adding a
locator to the objective was rejected twice over: inside the digest, a dead URL
would fork the objective and orphan every claim against it; outside it, it would
be unpinned attacker-controlled text that every verifying node fetches.

PARITY

This mirrors ``src/blobs.rs`` and must keep mirroring it. The two things that
have to agree exactly, or a Python node and a Rust node sharing a root will fail
to find each other's blobs:

* the address: bare lowercase hex, no ``sha256:`` prefix (that spelling belongs
  to record ids), of the file's exact bytes with no normalization;
* the layout: ``<root>/.proofwork/blobs/<address>``, flat.

``conformance/vectors.json`` pins both.

RULES

*Verify before write.* :meth:`BlobStore.put` hashes in memory and refuses a
mismatch before anything reaches the filesystem, so the store never holds a file
that could be executed under a name it does not hash to. The write is to a
temporary name and then renamed, so a crash mid-write cannot leave a short file
visible under a real address.

*A ceiling, checked before allocating.* :data:`MAX_BLOB_BYTES` bounds a blob
wherever it arrives, against a declared length where one exists.

*An address is 64 lowercase hex characters or it is not an address.*
:func:`is_address` is load-bearing rather than cosmetic: the address becomes a
filename and it arrives from an objective a stranger wrote. Without the check,
``"sha256": "../../../../etc/passwd"`` reaches ``os.path.join``.
"""
from __future__ import annotations

import hashlib
import os
from typing import Iterable

#: Where a store lives under a bundle root. Beside the pinned code rather than
#: beside the log, because the root is what a pinned path resolves against.
STORE_DIR = os.path.join(".proofwork", "blobs")

#: Largest blob this implementation will store, serve, or accept.
#:
#: A pinned checker is a source file; the shipped examples are a few kilobytes.
#: Deliberately not configurable: a node that raised it would accept blobs its
#: peers refuse, and that asymmetry is invisible until an objective becomes
#: unverifiable on half the network.
MAX_BLOB_BYTES = 1 << 20

_HEX = set("0123456789abcdef")


def address(data: bytes) -> str:
    """The content address of ``data``: bare lowercase hex, no prefix.

    Matches the spelling an objective's verifier block already uses. Two
    spellings for one hash would be two names for one blob, and each half of the
    network would fail to find the other half's copy.
    """
    return hashlib.sha256(data).hexdigest()


def is_address(text: object) -> bool:
    """Exactly 64 characters, each ``0-9a-f``.

    Uppercase is refused rather than folded: two spellings would be two files,
    and the objective declaring the uppercase form would be unverifiable on a
    node that stored the lowercase one.
    """
    return isinstance(text, str) and len(text) == 64 and set(text) <= _HEX


class BlobError(Exception):
    """Base for every refusal in this module."""


class NotAnAddress(BlobError):
    """Refused before it is allowed anywhere near a path join."""


class TooLarge(BlobError):
    """Over ``MAX_BLOB_BYTES``."""


class Mismatch(BlobError):
    """Bytes offered under an address they do not hash to. Never written."""


class Absent(BlobError):
    """No local copy.

    Distinct from every other class here because it is the one that must become
    UNAVAILABLE rather than a refusal: a blob this node does not have says
    nothing about anybody's artifact, and a peer can fix it by sending it.
    """


class BlobStore:
    """A directory of blobs, each filed under its own content address.

    Flat rather than sharded by hash prefix. Sharding keeps directory sizes down
    when there are millions of objects; a node holds one blob per distinct pinned
    checker it has heard of, and a two-level tree would buy nothing but a second
    place for the path arithmetic to go wrong.
    """

    def __init__(self, directory: str):
        self.directory = directory

    @classmethod
    def under(cls, root: str) -> "BlobStore":
        """The store belonging to a bundle root."""
        return cls(os.path.join(root, STORE_DIR))

    def path_of(self, addr: str) -> str:
        """Where a blob would live.

        Raises for anything that is not an address, so no caller can turn a
        hostile pin into a path outside the store.
        """
        if not is_address(addr):
            raise NotAnAddress(
                f"{addr!r} is not a content address (64 lowercase hex characters)"
            )
        return os.path.join(self.directory, addr)

    def holds(self, addr: str) -> bool:
        """Whether a copy is on disk. Existence only -- cheap enough to call once
        per pin per round.

        A file whose contents have rotted is reported as held here and caught by
        :meth:`read`, which is the only place the bytes are used.
        """
        try:
            return os.path.isfile(self.path_of(addr))
        except NotAnAddress:
            return False

    def read(self, wanted: str) -> bytes:
        """Read a blob, verifying it against its own filename.

        Not paranoia about our own writes: the store is an ordinary directory an
        operator can edit, a backup can restore badly, and a disk can corrupt. A
        file whose name is the hash of its contents and whose contents disagree
        is not a blob, so it is deleted rather than returned -- otherwise it
        shadows the real blob forever, since :meth:`holds` would go on reporting
        the address as held and the fetch that would replace it would never be
        attempted.
        """
        path = self.path_of(wanted)
        try:
            size = os.path.getsize(path)
        except FileNotFoundError:
            raise Absent(f"no local copy of verifier code {wanted}") from None
        # The declared length, before the allocation. A store entry is untrusted
        # input for the same reason a wire frame is.
        if size > MAX_BLOB_BYTES:
            raise TooLarge(f"verifier code is {size} bytes, limit is {MAX_BLOB_BYTES}")
        with open(path, "rb") as handle:
            data = handle.read()
        actual = address(data)
        if actual != wanted:
            try:
                os.remove(path)
            except OSError:
                pass
            raise Mismatch(f"verifier code hashes to {actual}, offered under {wanted}")
        return data

    def put(self, declared: str, data: bytes) -> str:
        """Store ``data`` under ``declared``, **verifying first**.

        Every path into the store goes through here, which is what makes "a
        received blob is checked against the pin before it is written anywhere it
        could be run" a property of the class rather than of each call site. The
        order is: shape of the address, then the ceiling, then the hash, then the
        filesystem. Nothing hostile reaches disk.
        """
        path = self.path_of(declared)
        if len(data) > MAX_BLOB_BYTES:
            raise TooLarge(f"verifier code is {len(data)} bytes, limit is {MAX_BLOB_BYTES}")
        actual = address(data)
        if actual != declared:
            raise Mismatch(f"verifier code hashes to {actual}, offered under {declared}")
        if os.path.isfile(path):
            return path
        os.makedirs(self.directory, mode=0o700, exist_ok=True)
        # Written under a temporary name and renamed, so a crash cannot leave a
        # truncated file visible under an address it does not hash to. The
        # temporary name carries the pid so two processes sharing a root do not
        # rename each other's partial writes into place.
        staging = os.path.join(self.directory, f".{declared}.{os.getpid()}.partial")
        try:
            # No execute bit: this file holds code a stranger sent, and the only
            # thing that should open it is the interpreter running the verifier.
            handle = os.open(staging, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o600)
            try:
                os.write(handle, data)
            finally:
                os.close(handle)
            os.replace(staging, path)
        except OSError as exc:
            try:
                os.remove(staging)
            except OSError:
                pass
            raise BlobError(f"blob store: {exc}") from exc
        return path

    def publish_file(self, path: str, declared: str) -> bool:
        """Publish a bundle file whose hash the objective declares.

        How code enters the network: the funder has the file on disk, and this
        makes their node able to serve it. ``False`` means the file is absent or
        does not match the pin -- not an error, because a node replaying somebody
        else's objectives has neither the bundle nor any obligation to.
        """
        if not is_address(declared):
            raise NotAnAddress(f"{declared!r} is not a content address")
        try:
            if not os.path.isfile(path) or os.path.getsize(path) > MAX_BLOB_BYTES:
                return False
            with open(path, "rb") as handle:
                data = handle.read()
        except OSError:
            return False
        if address(data) != declared:
            return False
        self.put(declared, data)
        return True

    def addresses(self) -> list[str]:
        """Every address held, sorted.

        An empty or absent directory is empty, not an error: a node that has
        never fetched anything is a normal node. Anything whose name is not an
        address is not a blob, whatever it contains.
        """
        try:
            names = os.listdir(self.directory)
        except OSError:
            return []
        return sorted(name for name in names if is_address(name))

    def retain(self, keep: Iterable[str]) -> int:
        """Drop every blob outside ``keep``, returning how many went.

        Collection is explicit -- a command an operator runs, never something a
        sync round does on its way past. A node cannot tell a blob nobody wants
        from a blob pinned by an objective it has not synced yet, and dropping
        the second kind on a timer would make a node stop being able to serve
        precisely the code its peers are about to ask it for.
        """
        keeping = set(keep)
        dropped = 0
        for held in self.addresses():
            if held in keeping:
                continue
            try:
                os.remove(self.path_of(held))
                dropped += 1
            except OSError:
                pass
        return dropped
